metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteEntityService.java

```diff
@@ -0,0 +1,254 @@
+import com.linkedin.metadata.query.filter.RelationshipDirection;
+import com.linkedin.metadata.run.DeleteReferencesResponse;
+import com.linkedin.metadata.run.RelatedAspect;
+import com.linkedin.metadata.run.RelatedAspectArray;
+import com.linkedin.metadata.utils.GenericRecordUtils;
+import com.linkedin.mxe.MetadataChangeProposal;
+import com.linkedin.util.Pair;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections.CollectionUtils;
+
+import static com.linkedin.metadata.search.utils.QueryUtils.*;
+
+
+@Slf4j
+@RequiredArgsConstructor
+public class DeleteEntityService {
+
+  private final EntityService _entityService;
+  private final GraphService _graphService;
+
+  private static final Integer ELASTIC_BATCH_DELETE_SLEEP_SEC = 5;
+
+  /**
+   * Public endpoint that deletes references to a given urn across DataHub's metadata graph. This is the entrypoint for
+   * addressing dangling pointers whenever a user deletes some entity.
+   *
+   * @param urn     The urn for which to delete references in DataHub's metadata graph.
+   * @param dryRun  Specifies if the delete logic should be executed to conclusion or if the caller simply wants a
+   *                preview of the response.
+   * @return  A {@link DeleteReferencesResponse} instance detailing the response of deleting references to the provided
+   * urn.
+   */
+  public DeleteReferencesResponse deleteReferencesTo(final Urn urn, final boolean dryRun) {
+    final DeleteReferencesResponse result = new DeleteReferencesResponse();
+    RelatedEntitiesResult relatedEntities =
+        _graphService.findRelatedEntities(null, newFilter("urn", urn.toString()), null,
+            EMPTY_FILTER,
+            ImmutableList.of(),
+            newRelationshipFilter(EMPTY_FILTER, RelationshipDirection.INCOMING), 0, 10000);
+
+    final List<RelatedAspect> relatedAspects = relatedEntities.getEntities().stream().map(relatedEntity -> {
+          final Urn relatedUrn = UrnUtils.getUrn(relatedEntity.getUrn());
+          final String relationType = relatedEntity.getRelationshipType();
+          final String relatedEntityName = relatedUrn.getEntityType();
+          final EntitySpec relatedEntitySpec = _entityService.getEntityRegistry().getEntitySpec(relatedEntityName);
+
+          return getAspectReferringTo(urn, relatedUrn, relationType, relatedEntitySpec).map(pair -> {
+            final RelatedAspect relatedAspect = new RelatedAspect();
+            relatedAspect.setEntity(relatedUrn);
+            relatedAspect.setRelationship(relationType);
+            relatedAspect.setAspect(pair.getSecond().getName());
+            return relatedAspect;
+          });
+        }).filter(java.util.Optional::isPresent)
```

Comment:
When is it not present?

Reply:
I see - it returns an Optional<Pair> - This data structure is super non friendly to read. Let's think about how to simplify. 

For now, can you rename pair to maybePair - this is the naming convention for Optionals in general